# Проект по SQL

Наша компания выкупила крупный сервис для чтения книг по подписке. Задача — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных:**

Таблица `books`
<br>Содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors`
<br>Содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers`
<br>Содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица `ratings`
<br>Содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews`
<br>Содержит данные о пользовательских обзорах:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

**Задачи:**
- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Загрузка и исследование данных

In [1]:
#!/usr/bin/python
# -*- coding: utf-8 -*-

import pandas as pd
from sqlalchemy import create_engine
import datetime as dt

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                  db_config['pwd'], 
                                                  db_config['host'], 
                                                  db_config['port'], 
                                                  db_config['db']) 

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    data_inspection = ''' SELECT * 
                          FROM {}; 
                      '''.format(table) 

    request = pd.io.sql.read_sql(data_inspection, con = engine) 
    print(f'Таблица - {table}')
    display(request.head())
    print(request.info())    
    print()

Таблица - books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None

Таблица - authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None

Таблица - publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None

Таблица - ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None

Таблица - reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None



Данные в порядке, пропусков нет. Приступим к выполнению задач.

# Задание 1.
Посчитаем, сколько книг вышло после 1 января 2000 года.

In [4]:
number_of_books = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 
request_1 = pd.io.sql.read_sql(number_of_books, con = engine)

print(f'После 1 января 2000 года вышло {request_1.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


# Задание 2.
Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [5]:
avg_rating = ''' 
SELECT 
      books.title AS title,
      books.book_id AS id,
      SubRev.n_rev AS number_of_reviews,
      SubRat.avg_rat AS avg_rating
FROM books LEFT JOIN 
                    (SELECT
                           book_id,
                           ROUND (AVG(rating), 2) AS avg_rat
                     FROM ratings
                     GROUP BY book_id
                    ) AS SubRat ON SubRat.book_id = books.book_id
           LEFT JOIN
                    (SELECT 
                           book_id,
                           COUNT(review_id) AS n_rev
                     FROM reviews
                     GROUP BY book_id
                    ) AS SubRev ON SubRev.book_id = books.book_id;               
''' 

request_2 = pd.io.sql.read_sql(avg_rating, con = engine)
request_2

title   id  \
0            The Body in the Library (Miss Marple  #3)  652   
1                                            Galápagos  273   
2                             A Tree Grows in Brooklyn   51   
3    Undaunted Courage: The Pioneering First Missio...  951   
4                                          The Prophet  839   
..                                                 ...  ...   
995                                Alice in Wonderland   64   
996          A Woman of Substance (Emma Harte Saga #1)   55   
997                                          Christine  148   
998  The Magicians' Guild (Black Magician Trilogy  #1)  790   
999                           The Plot Against America  828   

     number_of_reviews  avg_rating  
0                  2.0        4.50  
1                  2.0        4.50  
2                  5.0        4.25  
3                  2.0        4.00  
4                  4.0        4.29  
..                 ...         ...  
995                4.0        4.23  
996                2.0        5.00  
997                3.0        3.43  
998                2.0        3.50  
999                2.0        3.00  

[1000 rows x 4 columns]

Для каждой книги успешно посчитали количество обзоров и среднюю оценку.

# Задание 3.
Определим издательство, которое выпустило наибольшее число книг толще 50 страниц.

In [6]:
publisher = ''' 
SELECT
      pb.publisher,
      pb.n_books
FROM
    (SELECT 
           books.publisher_id AS publisher_id,
           COUNT(books.book_id) AS n_books,
           publishers.publisher AS publisher 
     FROM books LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id
     WHERE num_pages>50
     GROUP BY books.publisher_id,
           publishers.publisher
     ORDER BY n_books DESC) AS pb
LIMIT 1; 
''' 

request_3 = pd.io.sql.read_sql(publisher, con = engine)
#request_3
print(f'Издательство {request_3.iloc[0,0]} опубликовало наибольшее число книг - {request_3.iloc[0,1]} книги толще 50 страниц.')

Издательство Penguin Books опубликовало наибольшее число книг - 42 книги толще 50 страниц.


# Задание 4.
Определим автора с самой высокой средней оценкой книг, будем учитывать только книги с 50 и более оценками.

In [7]:
author = '''
SELECT author,
       ROUND(AVG(rating), 2) as avg_rating
FROM authors AS a INNER JOIN books AS b ON b.author_id=a.author_id
                  INNER JOIN ratings AS r ON r.book_id=b.book_id
WHERE b.book_id IN (SELECT  book_id
                    FROM ratings
                    GROUP BY book_id
                    HAVING COUNT(rating) > 50)
GROUP BY author
ORDER BY AVG(rating) DESC
LIMIT 1;
'''

request_4 = pd.io.sql.read_sql(author, con = engine) 
#request_4
print(f'Автор с самой высокой средней оценкой книгц - {request_4.iloc[0,0]}, средняя оценка книг {request_4.iloc[0,1]}')

Автор с самой высокой средней оценкой книгц - J.K. Rowling/Mary GrandPré, средняя оценка книг 4.29


# Задание 5.
Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [8]:
number_of_reviews = '''
WITH temp AS 
     (SELECT username,
            COUNT(review_id) AS review_cnt
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating) > 50)
      GROUP BY username)
    
SELECT ROUND(AVG(review_cnt), 1)
FROM temp;
'''

request_5 = pd.io.sql.read_sql(number_of_reviews, con = engine) 
print(f'Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - {request_5.iloc[0,0]}')

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3


# Выводы
По результатам исследования:
- После 1 января 2000 года вышло 819 книг;
- Для каждой книги успешно посчитали количество обзоров и среднюю оценку;
- Издательство Penguin Books опубликовало наибольшее число книг - 42 книги толще 50 страниц;
- Автор с самой высокой средней оценкой книгц - J.K. Rowling/Mary GrandPré, средняя оценка книг 4.29;
- Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3.